File: src/main/java/top/criwits/scaleda/verilog/psi/references/NamedPortConnectionReference.kt
//package top.criwits.scaleda.verilog.psi.references
//
//import com.intellij.psi.PsiElementResolveResult
//import com.intellij.psi.PsiReferenceBase
//import com.intellij.psi.ResolveResult
//import com.intellij.psi.util.PsiTreeUtil
//import top.criwits.scaleda.verilog.psi.factory.nodes.ModuleDeclarationPsiNode
//import top.criwits.scaleda.verilog.psi.factory.nodes.ModuleInstantiationPsiNode
//import top.criwits.scaleda.verilog.psi.factory.nodes.NamedPortConnectionPsiNode
//
//class NamedPortConnectionReferenceOld(element: NamedPortConnectionPsiNode) :
//    PsiReferenceBase.Poly<NamedPortConnectionPsiNode>(
//        element,
//        element.getHoldPsiNodeRelativeTextRange(),
//        true
//    ) {
//
//    override fun multiResolve(incompleteCode: Boolean): Array<ResolveResult> {
//        val moduleInstantiationPsiNode = PsiTreeUtil
//            .getParentOfType(myElement, ModuleInstantiationPsiNode::class.java)
//            ?: return emptyArray()
//        return (moduleInstantiationPsiNode.reference as? Poly<*> ?: return emptyArray())
//            .multiResolve(incompleteCode)
//            .map { it.element }
//            .filterIsInstance(ModuleDeclarationPsiNode::class.java)
////                .flatMap { it.availableNamedElements }
//            .filter { it.name == myElement.getHoldPsiNode()?.name }
//            .map { PsiElementResolveResult(it) }
//            .toTypedArray()
//    }
//
//    override fun getVariants(): Array<Any> {
//        return emptyArray()
//    }
//
//}

File: src/main/scala/top/criwits/scaleda/idea/verilog/references/NamedPortConnectionReference.scala
package top.criwits.scaleda
package idea.verilog.references

import verilog.psi.factory.nodes.{ModuleDeclarationPsiNode}

import com.intellij.psi.util.PsiTreeUtil
import com.intellij.psi.{PsiElementResolveResult, PsiReferenceBase, ResolveResult}
import top.criwits.scaleda.idea.verilog.psi.factory.nodes.NamedPortConnectionPsiNode

class NamedPortConnectionReference(element: NamedPortConnectionPsiNode)
  extends PsiReferenceBase.Poly[NamedPortConnectionPsiNode](
    element,
    element.getHoldPsiNodeRelativeTextRange,
    true
  ) {
  override def multiResolve(incompleteCode: Boolean): Array[ResolveResult] = {
    val moduleInstantiationPsiNode = PsiTreeUtil.getParentOfType(myElement, classOf[ModuleInstantiationPsiNode])
    moduleInstantiationPsiNode.getReference.multiResolve(incompleteCode)
      .map(it => it.getElement)
      .filter(it => it.isInstanceOf[ModuleDeclarationPsiNode])
      .map(it => it.asInstanceOf[ModuleDeclarationPsiNode])
      // .flatMap
      .filter(it => it.getName == myElement.getHoldPsiNode.getName)
      .map(new PsiElementResolveResult(_))
      .toArray
  }

  override def getVariants = Array()
}
